tools/autograd/templates/python_variable_methods.cpp

```diff
@@ -309,40 +309,6 @@ static PyObject * THPVariable_invert(PyObject* self, PyObject* args) {
   END_HANDLE_TH_ERRORS
 }
 
-static PyObject * THPVariable_cpu(PyObject* self, PyObject* args)
-{
-   HANDLE_TH_ERRORS
-   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
-   auto backend = self_.is_sparse() ? Backend::SparseCPU : Backend::CPU;
-   auto& type = self_.type().toBackend(backend);
-   return wrap(torch::utils::dispatch_type_conversion(self_, type));
-   END_HANDLE_TH_ERRORS
-}
-
-static PyObject * THPVariable_cuda(PyObject* self, PyObject* args, PyObject* kwargs)
-{
-  HANDLE_TH_ERRORS
-  static PythonArgParser parser({
-    "cuda(Device? device=None, bool non_blocking=False)",
-    "cuda(Device? device=None, bool async=False)|deprecated"
-  });
-  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
-  ParsedArgs<2> parsed_args;
-  auto r = parser.parse(args, kwargs, parsed_args);
-  auto backend = self_.is_sparse() ? at::kSparseCUDA : at::kCUDA;
-  auto& type = self_.type().toBackend(backend);
-  auto device_obj = r.device(0);
-  if (!r.isNone(0) && device_obj.is_cpu()) {
-    throw std::runtime_error("Invalid device, must be cuda device");
-  }
-  int32_t device_index = -1;
```

Comment:
If I remember correctly, this method and `dispatch_type_conversion` have some complex logic around `device_index` that we need to verify is preserved here.

Reply:
Yeah so if you follow the logic flow here, the semantics of this method are (cf https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/tensor_conversion_dispatch.cpp#L24):
1. If the device passed to `cuda(...)` supplies an index, the tensor will be allocated on that device,
2. Else, allocate on the device index of `self_`.

You are changing the semantics to:
1. If the device passed to `cuda(...)` supplies an index, the tensor will be allocated on that device,
2. Else, allocate on the current device.

Those two are not equivalent. A quick workaround would be to change the native function to (conceptually): 
```
return self.to({at::kCUDA, device_index == -1 ? (self.is_cuda() ? self.get_device() : -1) : device_index}, non_blocking);
```
but now what if the user actually did pass `Device(at::kCUDA, -1)`, and thus wanted the new tensor to be allocated on the default device? Then with the above scheme we'd be breaking it. That's why `dispatch_type_conversion` used an `at::optional<int32_t>` and then did funny logic around it.

Reply:
Yeah, this semantics problem kills this patch. So I guess we have to do `at::optional` here, which means we probably will have to solve `at::optional` before we can fix it. Blah.